cache/redis_client.go

```diff
@@ -0,0 +1,347 @@
+	f.IntVar(&c.ConnectionPoolSize, prefix+".connection-pool-size", 100, "Maximum number of connections in the pool.")
+	f.IntVar(&c.MinIdleConnections, prefix+".min-idle-connections", 10, "Minimum number of idle connections.")
+	f.DurationVar(&c.MaxConnectionAge, prefix+".max-connection-age", 0, "Close connections older than this duration. If the value is zero, then the pool does not close connections based on age.")
+	f.DurationVar(&c.IdleTimeout, prefix+".idle-timeout", time.Minute*5, "Amount of time after which client closes idle connections.")
+
+	f.IntVar(&c.MaxAsyncConcurrency, prefix+".max-async-concurrency", 50, "The maximum number of concurrent asynchronous operations can occur.")
+	f.IntVar(&c.MaxAsyncBufferSize, prefix+".max-async-buffer-size", 25000, "The maximum number of enqueued asynchronous operations allowed.")
+	f.IntVar(&c.MaxGetMultiConcurrency, prefix+".max-get-multi-concurrency", 100, "The maximum number of concurrent connections running get operations. If set to 0, concurrency is unlimited.")
+	f.IntVar(&c.GetMultiBatchSize, prefix+".max-get-multi-batch-size", 100, "The maximum size per batch for mget operations.")
+	f.IntVar(&c.MaxItemSize, prefix+".max-item-size", 16*1024*1024, "The maximum size of an item stored in Redis. Bigger items are not stored. If set to 0, no maximum size is enforced.")
+
+	f.BoolVar(&c.TLSEnabled, prefix+".tls-enabled", false, "Enable connecting to Redis with TLS.")
+	c.TLS.RegisterFlagsWithPrefix(prefix, f)
+}
+
+func (c *RedisClientConfig) Validate() error {
+	if c.Endpoint.String() == "" {
+		return errRedisConfigNoEndpoint
+	}
+	// Set async only available when MaxAsyncConcurrency > 0.
+	if c.MaxAsyncConcurrency <= 0 {
+		return errRedisMaxAsyncConcurrencyNotPositive
+	}
+	return nil
+}
+
+type redisClient struct {
+	redis.UniversalClient
+	config RedisClientConfig
+
+	// Queue used to enqueue async operations.
+	asyncQueue *asyncQueue
+
+	// getMultiGate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+
+	logger  log.Logger
+	metrics *clientMetrics
+}
+
+// NewRedisClient makes a new RedisClient.
+func NewRedisClient(logger log.Logger, name string, config RedisClientConfig, reg prometheus.Registerer) (RemoteCacheClient, error) {
+	opts := &redis.UniversalOptions{
+		Addrs:        strings.Split(config.Endpoint.String(), ","),
+		Username:     config.Username,
+		Password:     config.Password.String(),
+		DB:           config.DB,
+		MasterName:   config.MasterName,
+		DialTimeout:  config.DialTimeout,
+		ReadTimeout:  config.ReadTimeout,
+		WriteTimeout: config.WriteTimeout,
+		PoolSize:     config.ConnectionPoolSize,
+		MinIdleConns: config.MinIdleConnections,
+		MaxConnAge:   config.MaxConnectionAge,
+		IdleTimeout:  config.IdleTimeout,
+	}
+
+	if config.TLSEnabled {
+		tlsConfig := config.TLS
+
+		tlsClientConfig, err := newClientConfig(logger, tlsConfig.CertPath, tlsConfig.KeyPath,
```

Comment:
I don't think we should use `newClientConfig()` at all. Since we use dskit TLS config struct, why don't we just use `config.TLS.GetTLSConfig()`?